Upgrade logos dependency in lunarity

Our project has a dependency that does not compile on latest nightly

```
$ cargo +nightly check
   Compiling logos-derive v0.7.7
error[E0061]: this function takes 1 argument but 2 arguments were supplied
   --> /home/birchmd/.cargo/registry/src/github.com-1ecc6299db9ec823/logos-derive-0.7.7/src/lib.rs:55:20
    |
55  |             extras.insert(util::ident(&ext), |_| panic!("Only one #[extras] attribute can be declared."));
    |                    ^^^^^^ -----------------  ----------------------------------------------------------- supplied 2 arguments
    |                    |
    |                    expected 1 argument
    |
note: associated function defined here

error[E0061]: this function takes 1 argument but 2 arguments were supplied
   --> /home/birchmd/.cargo/registry/src/github.com-1ecc6299db9ec823/logos-derive-0.7.7/src/lib.rs:89:23
    |
89  |                 error.insert(variant, |_| panic!("Only one #[error] variant can be declared."));
    |                       ^^^^^^ -------  -------------------------------------------------------- supplied 2 arguments
    |                       |
    |                       expected 1 argument
    |
note: associated function defined here

error[E0061]: this function takes 1 argument but 2 arguments were supplied
   --> /home/birchmd/.cargo/registry/src/github.com-1ecc6299db9ec823/logos-derive-0.7.7/src/lib.rs:93:21
    |
93  |                 end.insert(variant, |_| panic!("Only one #[end] variant can be declared."));
    |                     ^^^^^^ -------  ------------------------------------------------------ supplied 2 arguments
    |                     |
    |                     expected 1 argument
    |
note: associated function defined here

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0061`.
error: could not compile `logos-derive`

To learn more, run the command again with --verbose.
```

This `logos-derive` dependency comes from our [dependency on `lunarity`](https://github.com/aurora-is-near/aurora-engine/blob/8b767e44a5378da090c3ceb9113af4d7378f9216/Cargo.toml#L49). Fortunately, we are already working off a fork which we control (it's on Illia's GitHub), and the latest version of `logos` (v0.12) _does_ compile with latest nightly. So we can resolve this issue by upgrading `logos` to the latest version in our fork of the `lunarity` project.

Unfortunately, this upgrade is non-trivial because there were [several breaking API changes](https://github.com/maciejhirsz/logos/releases/tag/v0.11.0) in `logos`, so some code in `lunarity` will need to be refactored to work with the latest version. This issue represents the work to do that refactoring and thus have a version of `lunarity` that compiles in latest nightly for our project to depend on.

This is probably a low priority item since it will likely not be blocking us until we decide to upgrade our `rust-toolchain` file to a later version of rust.

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -51,7 +51,7 @@ rlp = { version = "0.5.0", default-features = false }
 sha2 = { version = "0.9.3", default-features = false, optional = true }
 sha3 = { version = "0.9.1", default-features = false }
 wee_alloc = { version = "0.4.5", default-features = false }
-lunarity-lexer = { git = "https://github.com/ilblackdragon/lunarity", rev = "5201d9a76f7e491082b7f74af7e64049271e387f", default-features = false }
+logos = { version = "0.12", default-features = false, features = ["export_derive"] }
 ethabi = { git = "https://github.com/darwinia-network/ethabi", branch = "xavier-no-std", default-features = false }
 hex = { version = "0.4", default-features = false, features = ["alloc"] }
 byte-slice-cast = { version = "1.0", default-features = false }
@@ -73,7 +73,7 @@ git2 = "0.13"
 
 [features]
 default = ["sha2", "std"]
-std = ["borsh/std", "evm/std", "primitive-types/std", "rlp/std", "sha3/std", "ethabi/std", "lunarity-lexer/std", "bn/std"]
+std = ["borsh/std", "evm/std", "primitive-types/std", "rlp/std", "sha3/std", "ethabi/std", "logos/std", "bn/std"]
 testnet = []
 engine = []
 contract = ["engine"]

--- a/src/meta_parsing.rs
+++ b/src/meta_parsing.rs
@@ -1,6 +1,6 @@
 use borsh::BorshDeserialize;
 use ethabi::{encode, Token as ABIToken};
-use lunarity_lexer::{Lexer, Token};
+use logos::Logos;
 use rlp::{Decodable, DecoderError, Rlp};
 
 use crate::parameters::MetaCallArgs;
@@ -17,6 +17,70 @@ pub enum ParsingError {
 }
 
 pub type ParsingResult<T> = core::result::Result<T, ParsingError>;
+
+mod type_lexer {
+    use logos::{Lexer, Logos};
+
+    #[derive(Logos, Debug, PartialEq)]
+    pub(super) enum Token {
+        #[regex("byte|bytes[1-2][0-9]?|bytes3[0-2]?|bytes[4-9]", fixed_bytes_size)]
+        FixedBytes(u8),
+        #[regex("uint(8|16|24|32|40|48|56|64|72|80|88|96|104|112|120|128|136|144|152|160|168|176|184|192|200|208|216|224|232|240|248|256)?", |lex| fixed_int_size(lex, "uint"))]
+        Uint(usize),
+        #[regex("int(8|16|24|32|40|48|56|64|72|80|88|96|104|112|120|128|136|144|152|160|168|176|184|192|200|208|216|224|232|240|248|256)?", |lex| fixed_int_size(lex, "int"))]
+        Int(usize),
+        #[regex("bool")]
+        Bool,
+        #[regex("address")]
+        Address,
+        #[regex("bytes")]
+        Bytes,
+        #[regex("string")]
+        String,
+        #[regex("\\[[0-9]*\\]", reference_type_size)]
+        ReferenceType(Option<u64>),
+        #[regex("[a-zA-Z_$][a-zA-Z0-9_$]*")]
+        Identifier,
+
+        #[error]
+        Error,
+    }
+
+    fn fixed_bytes_size(lex: &mut Lexer<Token>) -> u8 {
+        let slice = lex.slice();
+
+        if slice == "byte" {
+            return 1;
+        }
+
+        let n = slice["bytes".len()..].parse();
+        n.ok().unwrap_or(1)
+    }
+
+    fn fixed_int_size(lex: &mut Lexer<Token>, prefix: &str) -> usize {
+        let slice = lex.slice();
+
+        if slice == prefix {
+            // the default int size is 32
+            return 32;
+        }
+
+        let n = slice[prefix.len()..].parse();
+        n.unwrap_or(32)
+    }
+
+    fn reference_type_size(lex: &mut Lexer<Token>) -> Option<u64> {
+        let slice = lex.slice();
+
+        if slice == "[]" {
+            return None;
+        }
+
+        let end_index = slice.len() - 1;
+        let n = slice[1..end_index].parse();
+        n.ok()
+    }
+}
 
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub enum ArgType {
@@ -38,71 +102,35 @@ pub enum ArgType {
 /// field_type: A single evm function arg type in string, without the argument name
 /// e.g. "bytes" "uint256[][3]" "CustomStructName"
 pub fn parse_type(field_type: &str) -> ParsingResult<ArgType> {
-    #[derive(PartialEq)]
-    enum State {
-        Open,
-        Close,
-    }
+    let mut lexer = type_lexer::Token::lexer(field_type);
+    let mut current_token = lexer.next();
+    let mut inner_type: Option<ArgType> = None;
 
-    let mut lexer = Lexer::new(field_type);
-    let mut token = None;
-    let mut state = State::Close;
-    let mut array_depth = 0;
-    let mut current_array_length: Option<u64> = None;
-
-    while lexer.token != Token::EndOfProgram {
-        let type_ = match lexer.token {
-            Token::Identifier => ArgType::Custom(lexer.slice().to_owned()),
-            Token::TypeByte => ArgType::Byte(lexer.extras.0),
-            Token::TypeBytes => ArgType::Bytes,
-            Token::TypeBool => ArgType::Bool,
-            Token::TypeUint => ArgType::Uint,
-            Token::TypeInt => ArgType::Int,
-            Token::TypeString => ArgType::String,
-            Token::TypeAddress => ArgType::Address,
-            Token::LiteralInteger => {
-                let length = lexer.slice();
-                current_array_length = Some(
-                    length
-                        .parse()
-                        .map_err(|_| ParsingError::InvalidMetaTransactionMethodName)?,
-                );
-                lexer.advance();
-                continue;
-            }
-            Token::BracketOpen if token.is_some() && state == State::Close => {
-                state = State::Open;
-                lexer.advance();
-                continue;
-            }
-            Token::BracketClose if array_depth < 10 => {
-                if state == State::Open && token.is_some() {
-                    let length = current_array_length.take();
-                    state = State::Close;
-                    token = Some(ArgType::Array {
-                        inner: Box::new(token.expect("if statement checks for some; qed")),
-                        length,
-                    });
-                    lexer.advance();
-                    array_depth += 1;
-                    continue;
-                } else {
-                    return Err(ParsingError::InvalidMetaTransactionMethodName);
-                }
-            }
-            Token::BracketClose if array_depth == 10 => {
-                return Err(ParsingError::InvalidMetaTransactionMethodName);
-            }
-            _ => {
-                return Err(ParsingError::InvalidMetaTransactionMethodName);
-            }
+    loop {
+        let typ = match current_token {
+            None => break,
+            Some(type_lexer::Token::Address) => ArgType::Address,
+            Some(type_lexer::Token::Bool) => ArgType::Bool,
+            Some(type_lexer::Token::String) => ArgType::String,
+            Some(type_lexer::Token::Bytes) => ArgType::Bytes,
+            Some(type_lexer::Token::Identifier) => ArgType::Custom(lexer.slice().to_owned()),
+            Some(type_lexer::Token::FixedBytes(size)) => ArgType::Byte(size),
+            Some(type_lexer::Token::Int(_)) => ArgType::Int,
+            Some(type_lexer::Token::Uint(_)) => ArgType::Uint,
+            Some(type_lexer::Token::ReferenceType(length)) => match inner_type {
+                None => return Err(ParsingError::ArgumentParseError),
+                Some(t) => ArgType::Array {
+                    length,
+                    inner: Box::new(t),
+                },
+            },
+            Some(type_lexer::Token::Error) => return Err(ParsingError::ArgumentParseError),
         };
-
-        token = Some(type_);
-        lexer.advance();
+        inner_type = Some(typ);
+        current_token = lexer.next();
     }
 
-    token.ok_or(ParsingError::InvalidMetaTransactionMethodName)
+    inner_type.ok_or(ParsingError::ArgumentParseError)
 }
 
 /// NEAR's domainSeparator

--- a/src/meta_parsing.rs
+++ b/src/meta_parsing.rs
@@ -603,3 +603,121 @@ pub fn parse_meta_call(
         Err(_) => Err(ParsingError::InvalidEcRecoverSignature),
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::ArgType;
+    use rand::Rng;
+
+    #[test]
+    fn test_parse_type() {
+        // # atomic types
+
+        // ## bytesN
+        for n in 1..=32 {
+            let s = format!("bytes{}", n);
+            assert_arg_type(&s, ArgType::Byte(n));
+        }
+        assert_arg_type("byte", ArgType::Byte(1));
+
+        // ## uintN
+        for n in 1..=32 {
+            let s = format!("uint{}", 8 * n);
+            assert_arg_type(&s, ArgType::Uint);
+        }
+        assert_arg_type("uint", ArgType::Uint);
+
+        // ## intN
+        for n in 1..=32 {
+            let s = format!("int{}", 8 * n);
+            assert_arg_type(&s, ArgType::Int);
+        }
+        assert_arg_type("int", ArgType::Int);
+
+        // ## bool
+        assert_arg_type("bool", ArgType::Bool);
+
+        // ## address
+        assert_arg_type("address", ArgType::Address);
+
+        // ## custom
+        let mut rng = rand::thread_rng();
+        for _ in 0..u8::MAX {
+            let name = rand_identifier(&mut rng);
+            assert_arg_type(&name, ArgType::Custom(name.clone()));
+        }
+
+        // # dynamic types
+
+        // ## bytes
+        assert_arg_type("bytes", ArgType::Bytes);
+
+        // ## string
+        assert_arg_type("string", ArgType::String);
+
+        // # arrays
+        let inner_types: Vec<String> = (1..=32)
+            .map(|n| format!("bytes{}", n))
+            .chain((1..=32).map(|n| format!("uint{}", 8 * n)))
+            .chain((1..=32).map(|n| format!("int{}", 8 * n)))
+            .chain(std::iter::once("bool".to_string()))
+            .chain(std::iter::once("address".to_string()))
+            .chain(std::iter::once(rand_identifier(&mut rng)))
+            .chain(std::iter::once("bytes".to_string()))
+            .chain(std::iter::once("string".to_string()))
+            .collect();
+        for t in inner_types {
+            let inner_type = super::parse_type(&t).ok().unwrap();
+            let size: Option<u8> = rng.gen();
+
+            // single array
+            let single_array_string = create_array_type_string(&t, size);
+            let expected = ArgType::Array {
+                length: size.map(|x| x as u64),
+                inner: Box::new(inner_type),
+            };
+            assert_arg_type(&single_array_string, expected.clone());
+
+            // nested array
+            let inner_type = expected;
+            let size: Option<u8> = rng.gen();
+            let nested_array_string = create_array_type_string(&single_array_string, size);
+            let expected = ArgType::Array {
+                length: size.map(|x| x as u64),
+                inner: Box::new(inner_type),
+            };
+            assert_arg_type(&nested_array_string, expected);
+        }
+
+        // # errors
+        // ## only numbers
+        super::parse_type("27182818").unwrap_err();
+        // ## invalid characters
+        super::parse_type("Some.InvalidType").unwrap_err();
+        super::parse_type("Some::NotType").unwrap_err();
+        super::parse_type("*AThing*").unwrap_err();
+    }
+
+    fn create_array_type_string(inner_type: &str, size: Option<u8>) -> String {
+        format!(
+            "{}[{}]",
+            inner_type,
+            size.map(|x| x.to_string()).unwrap_or(String::new())
+        )
+    }
+
+    fn assert_arg_type(s: &str, expected: ArgType) {
+        assert_eq!(super::parse_type(s).ok().unwrap(), expected);
+    }
+
+    fn rand_identifier<T: Rng>(rng: &mut T) -> String {
+        use rand::distributions::Alphanumeric;
+        use rand::seq::IteratorRandom;
+
+        // The first character must be a letter, so we sample that separately.
+        let first_char = ('a'..='z').chain('A'..='Z').choose(rng).unwrap();
+        let other_letters = (0..7).map(|_| char::from(rng.sample(Alphanumeric)));
+
+        std::iter::once(first_char).chain(other_letters).collect()
+    }
+}

